blog/_posts/2017-02-23-scalafix-v0.3.md

```diff
@@ -0,0 +1,141 @@
+The tree is backed by the scala.meta semantic API, so the rewrite is able to query for compiler information such as symbols.
+To demonstrate how rewrites are implemented with scalafix v0.3, let's step through an example use-case.
+
+## Example: Xor to Either
+
+The functional programming library [cats][cats] migrated recently from its `Xor` data type to `Either` from the standard library.
+It requires a few mechanical steps to migrate code to use `Either` instead of `Xor`.
+For example, below is a diff that's taken from [circe][xor-circe]'s migration to `Either`.
+
+```diff
+-final def either: Xor[HCursor, HCursor] = if (succeeded) Xor.right(any) else Xor.left(any)
++final def either: Either[HCursor, HCursor] = if (succeeded) Right(any) else Left(any)
+```
+
+For this particular rewrite, we are able to get away with only using tree patches.
+`Replace` is one tree patch that can be used to replace usage of a reference such as a type, term or a static method.
+
+```scala
+Replace(Symbol("_root_.cats.data.Xor."), q"Either")
+Replace(Symbol("_root_.cats.data.Xor.Left."), q"Left")
+Replace(Symbol("_root_.cats.data.Xor.Right."), q"Right")
+```
+
+The `Symbol(_root_...data.Xor)` part is the scala.meta symbol referencing the class definition of `cats.data.Xor`.
+As we saw in the `either: Xor[HCursor` diff above, references to `Xor` should become `Either` after the rewrite.
+
+Symbols are normalized by default, so the `cats.data.Xor.Right` replace patch will handle the `Xor.Right` type, `Xor.Right` companion object as well as the `Right.apply` constructor method.
+
+To introduce new imports, it's possible to pass in `additionalImports`
+
+```scala
+Replace(Symbol("_root_.cats.data.XorT."), q"EitherT",
+        additionalImports = List(importer"cats.data.EitherT")),
+```
+
+Imports can be removed with the `RemoveGlobalImport` patch
+
+```scala
+RemoveGlobalImport(importer"cats.data.Xor")
+```
+
+Nothing happens if the import does not appear in the source file.
+Likewise, it's OK to add the same import twice, scalafix will de-duplicate it.
+
+The full `Xor` to `Either` scalafix rewrite can be found [here][xor2either] and its accompanying test
+suite [here][xor2either-test].
+
+## Try it out!
+You can use scalafix both as a library and a tool.
+
+The recommended way to use scalafix as a library is with the [`sbt-scalahost`][sbt-scalahost] plugin.
+By using scalafix as a library, you have full control of how, where and when to run rewrites.
+
+The recommended way to use scalafix as tool is the [`sbt-scalafix`][sbt-scalafix] plugin.
+It's possible define custom tree patches in the [`.scalafix.conf` configuration file][config-patches].
+
+The long-term mission of scalafix is to help automate the migration of deprecated Scala 2.x features to Dotty.
+However, as I have hopefully demonstrated in this post, scalafix can be used for more than just migrating between Scala versions.
+Scalafix can also be used for ad-hoc library and application migrations.
```

Comment:
I avoid it/this like the plague! I agree it may sound repetitive but at least there's no ambiguity! Every sentence can be read in isolation :)